File: 2020/25.jl
include("utils.jl")
using .AoCUtils: readInput

function loadInput()
    filename = "input_25.txt"
    lines = readInput(filename)

    return parse.(Int, lines)
end

function calc_loop_size(public_key::Int)
    subject_number = 7
    n = 1
    i = 0
    while true
        n = (n * subject_number) % 20201227
        if n == public_key
            return i+1
        end
        i += 1
    end
end

function calc_encryption_key(subject_number::Int, loop_size::Int)
    n = 1
    for i in 1:loop_size
        n = (n * subject_number) % 20201227
    end
    return n
end

function part1(public_key_card::Int, public_key_door::Int)
    loop_size = calc_loop_size(public_key_card)
    enc_key = calc_encryption_key(public_key_door, loop_size)

    println("🎄 Part 1: $(enc_key)")
end

println("⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄")
println("⛄        Day 25        ⛄")
println("⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄")

public_key_card, public_key_door = loadInput()
@time part1(public_key_card, public_key_door)

File: 2020/25.py
import time
from utils import readInput

def loadInput():
    lines = readInput("input_25.txt")
    return [int(n) for n in lines]
 
def transform(subject_number, target):
    ls = 0
    s = 1
    while True:
        s *= subject_number
        s = s%20201227
        if s == target:
            return ls+1
        ls += 1

def encr_key(loop_size, subject_number):
    s = 1
    for _ in range(loop_size):
        s *= subject_number
        s = s % 20201227
    return s

def part1(card_pub_key, door_pub_key):
    card_loop_size = transform(7, card_pub_key)
    print ("🎄 Part 1: {}".format(encr_key(card_loop_size, door_pub_key)))

print("⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄")
print("⛄        Day 25        ⛄")
print("⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄")

card_pub_key, door_pub_key = loadInput()

t0 = time.time()
part1(card_pub_key, door_pub_key)
print ("Time: {:.5f}".format(time.time()-t0))

